File: solution/840. Magic Squares In Grid/840.cpp
class Solution {
public:
    int numMagicSquaresInside(vector<vector<int>>& grid) {
        int m = grid.size(), n = grid[0].size();
        
        int target = 0;
        for (int i = 1; i <= 9; ++i)
            target |= (1 << i);
        
        int res = 0;
        for (int i = 0; i + 3 <= m; ++i) {
            for (int j = 0; j + 3 <= n; ++j) {
                // check 1 to 9
                int mask = 0;
                for (int y = i; y < i + 3; ++y) {
                    for (int x = j; x < j + 3; ++x) {
                        mask |= (1 << grid[y][x]);
                    }
                }
                if (mask != target) continue;
                
                // check sum of each row
                int same = true;
                int sum = grid[i][j] + grid[i][j + 1] + grid[i][j + 2];
                for (int y = i + 1; y < i + 3 && same; ++y) {
                    if (grid[y][j] + grid[y][j + 1] + grid[y][j + 2] != sum)
                        same = false;
                }
                // check sum of each col
                for (int x = j; x < j + 3 && same; ++x) {
                    if (grid[i][x] + grid[i + 1][x] + grid[i + 2][x] != sum)
                        same = false;
                }
                // check sum of diagonals
                same &= grid[i][j] + grid[i + 1][j + 1] + grid[i + 2][j + 2] == sum && grid[i + 2][j] + grid[i + 1][j + 1] + grid[i][j + 2] == sum;
                res += same;
            }
        }
        return res;
    }
};

File: solution/840. Magic Squares In Grid/840.py
class Solution:
    def numMagicSquaresInside(self, grid: List[List[int]]) -> int:
        m, n = len(grid), len(grid[0])
    
        target = 0
        for i in range(1, 10):
            target |= (1 << i)

        def is_magic(i, j, target):
            for y in range(i, i + 3):
                for x in range(j, j + 3):
                    target ^= (1 << grid[y][x])
            if target != 0:
                return False

            s = grid[i][j] + grid[i+1][j+1] + grid[i+2][j+2]
            if grid[i][j+2] + grid[i+1][j+1] + grid[i+2][j] != s:
                return False

            for y in range(i, i + 3):
                if grid[y][j] + grid[y][j+1] + grid[y][j+2] != s:
                    return False
            for x in range(j, j + 3):
                if grid[i][x] + grid[i+1][x] + grid[i+2][x] != s:
                    return False
            return True
    
        res = 0
        for i in range(m - 2):
            for j in range(n - 2):
                if is_magic(i, j, target):
                    res += 1
        return res
    
